File: androidApp/src/main/java/com/bill/note_taking_app/android/note_detail/NoteDetailViewModel.kt
package com.bill.note_taking_app.android.note_detail

import androidx.lifecycle.SavedStateHandle
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.bill.note_taking_app.domain.note.Note
import com.bill.note_taking_app.domain.note.NoteDataSource
import com.bill.note_taking_app.domain.time.DateTimeUtil
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.*
import kotlinx.coroutines.launch
import javax.inject.Inject

@HiltViewModel
class NoteDetailViewModel @Inject constructor(
  private val noteDataSource: NoteDataSource,
  private val savedStateHandle: SavedStateHandle
) : ViewModel() {
  private val noteTitle = savedStateHandle.getStateFlow("noteTitle", "")
  private val isNoteTitleFocused = savedStateHandle.getStateFlow(
    "isNoteTitleHintVisible",
    false
  )
  private val noteContent = savedStateHandle.getStateFlow("noteContent", "")
  private val isNoteContentFocused = savedStateHandle.getStateFlow(
    "isNoteContentHintVisible",
    false
  )
  private val noteColour = savedStateHandle.getStateFlow(
    "noteColour",
    Note.generateRandomColour()
  )

  val state = combine(
    noteTitle,
    isNoteTitleFocused,
    noteContent,
    isNoteContentFocused,
    noteColour
  ) { title, isTitleFocused, content, isContentFocused, colour ->
    NoteDetailState(
      noteTitle = title,
      isNoteTitleHintVisible = title.isEmpty() && !isTitleFocused,
      noteContent = content,
      isNoteContentHintVisible = content.isEmpty() && !isContentFocused,
      noteColour = colour
    )
  }.stateIn(viewModelScope, SharingStarted.WhileSubscribed(5000), NoteDetailState())

  private var _hasNoteBeenSaved = MutableStateFlow(false)
  val hasNoteBeenSaved = _hasNoteBeenSaved.asStateFlow()

  private var existingNoteId: Long? = null

  init {
    savedStateHandle.get<Long>("noteId")?.let { existingNoteId ->
      if (existingNoteId == -1L) return@let

      this.existingNoteId = existingNoteId

      viewModelScope.launch {
        noteDataSource.getNoteById(existingNoteId)?.let { note ->
          savedStateHandle["noteTitle"] = note.title
          savedStateHandle["noteContent"] = note.content
          savedStateHandle["noteColour"] = note.colourHex
        }
      }
    }
  }

  fun onNoteTitleChange(text: String) {
    savedStateHandle["noteTitle"] = text
  }

  fun onNoteContentChange(text: String) {
    savedStateHandle["noteContent"] = text
  }

  fun onTitleFocusChange(isFocused: Boolean) {
    savedStateHandle["isNoteTitleHintVisible"] = isFocused
  }

  fun onContentFocusChange(isFocused: Boolean) {
    savedStateHandle["isNoteContentHintVisible"] = isFocused
  }

  fun saveNote() {
    viewModelScope.launch {
      noteDataSource.insertNote(
        Note(
          id = existingNoteId,
          title = noteTitle.value,
          content = noteContent.value,
          colourHex = noteColour.value,
          created = DateTimeUtil.now()
        )
      )
      _hasNoteBeenSaved.value = true
    }
  }
}

File: iosApp/iosApp/note_detail/NoteDetailViewModel.swift
//
//  NoteDetailViewModel.swift
//  iosApp
//
//  Created by Bill on 04/01/2023.
//  Copyright © 2023 orgName. All rights reserved.
//

import Foundation
import shared


extension NoteDetailScreen {
    @MainActor class NoteDetailViewModel: ObservableObject {
        private var noteDataSource: NoteDataSource?
        private var noteId: Int64? = nil
        @Published var noteTitle: String = ""
        @Published var noteContent: String = ""
        private(set) var noteColour = Note.Companion().generateRandomColour()
        
        init(noteDataSource: NoteDataSource? = nil) {
            self.noteDataSource = noteDataSource
        }
        
        func loadNoteIfExists(id: Int64?) {
            if id != nil {
                self.noteId = id
                noteDataSource?.getNoteById(id: id!, completionHandler: { note, error in
                    self.noteTitle = note?.title ?? ""
                    self.noteContent = note?.content ?? ""
                    self.noteColour = note?.colourHex ?? Note.Companion().generateRandomColour()
                })
            }
        }
        
        func saveNote(onSaved: @escaping () -> Void) {
            noteDataSource?.insertNote(note: Note(
                id: noteId == nil ? nil : KotlinLong(value: noteId!),
                title: noteTitle,
                content: noteContent,
                colourHex: noteColour,
                created: DateTimeUtil().now()
            ),
                completionHandler: { error in
                onSaved()
            })
        }
        
        func setParameters(noteDataSource: NoteDataSource, noteId: Int64?) {
            self.noteDataSource = noteDataSource
            loadNoteIfExists(id: noteId)
        }
    }
}
